src/linker/Linker/TypeReferenceExtensions.cs

```diff
@@ -12,46 +13,94 @@ public static string GetDisplayName (this TypeReference type)
 			if (type == null)
 				return string.Empty;
 
-			var builder = new System.Text.StringBuilder ();
-			if (type.DeclaringType != null)
-				builder.Append (type.DeclaringType.GetDisplayName ()).Append (".");
-
-			if (type.HasGenericParameters) {
-				builder.Append (type.Name.Substring (0, type.Name.IndexOf ('`')));
-				builder.Append ('<');
-				for (int i = 0; i < type.GenericParameters.Count - 1; i++)
-					builder.Append ($"{type.GenericParameters[i]},");
-
-				builder.Append ($"{type.GenericParameters[type.GenericParameters.Count - 1]}>");
-			} else if (type is ArrayType arrayType) {
-				void parseArrayDimensions (ArrayType at)
-				{
-					builder.Append ('[');
-					for (int i = 0; i < at.Dimensions.Count - 1; i++)
-						builder.Append (',');
-
-					builder.Append (']');
+			var sb = new StringBuilder ();
+			Stack<TypeReference> genericArguments = null;
+			while (true) {
+				switch (type) {
+				case ArrayType arrayType:
+					ParseArrayType (arrayType, sb);
+					break;
+				case GenericInstanceType genericInstanceType:
+					int arguments = int.Parse (genericInstanceType.Name.Substring (genericInstanceType.Name.IndexOf ('`') + 1));
+					genericArguments = new Stack<TypeReference> (genericInstanceType.GenericArguments);
```

Comment:
I must admit I don't really understand why we're doing this. The number of generic arguments can be determined by looking at the `GenericArguments` collection. And why do we need to allocate stack to reverse the order, only to then use prepend semantics... it feels backwards to me.